src/backend/mpi/swdsm.cpp

```diff
@@ -236,15 +236,28 @@ inline std::size_t align_backwards(std::size_t offset, std::size_t size) {
 	return (offset / size) * size;
 }
 
-void handler(int sig, siginfo_t *si, void *unused){
+void handler(int sig, siginfo_t *si, void *context){
 	UNUSED_PARAM(sig);
-	UNUSED_PARAM(unused);
 	double t1 = MPI_Wtime();
 	unsigned long tag;
 	argo_byte owner,state;
+
 	/* compute offset in distributed memory in bytes, always positive */
 	const std::size_t access_offset = static_cast<char*>(si->si_addr) - static_cast<char*>(startAddr);
 
+
+#ifdef REG_ERR
+	/* On x86, get and decode the error number from the context */
+	const ucontext_t* ctx = static_cast<ucontext_t*>(context);
+	auto err_num = ctx->uc_mcontext.gregs[REG_ERR];
+	assert(err_num & X86_PF_USER);	//Assert signal from user space
+	assert(err_num < X86_PF_RSVD); 	//Assert signal is from read or write access
+	/* This could be further decoded by using X86_PF_PROT to detect
+	 * whether the fault originated from no page found (0) or from
+	 * a protection fault (1), but is not needed for this purpose. */
+	bool write_miss = err_num & X86_PF_WRITE; //True for write, false for read
```

Comment:
would the later code be easier if we set this variable outside the `#ifdef`, but replicate the "old" behaviour (first=read, second=write) in its value?

Reply:
I have re-written the code slightly according to this suggestion. Write misses are now classified as `read`, `write` or `undefined` with the latter as default, and we make optimizations in the case where `read` or `write` miss knowledge can be used without checking for the definition of `REG_ERR`.